add porting guide for bare vars in conditionals

s-hertel opened the pull request:
##### SUMMARY
Fixes #54996

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
porting guide

s-hertel committed:
add porting guide for conditionals

s-hertel committed:
extra word

s-hertel committed:
remove extra whitespace

acozine reviewed (changes requested):
Please don't merge - I'm going to edit this a bit.

acozine committed:
revises bare variables entry

sanding

better consistency

first review feedback

more smooth edges

strings can be emtpy

acozine committed:
reorder based on sivel's feedback

acozine committed:
better headings

samccann commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -62,12 +63,94 @@ Beginning in version 2.8, Ansible will warn if a module expects a string, but a
+------------------------------
+
+In Ansible 2.7 and earlier, top-level variables sometimes treated boolean strings as if they were boolean values. This led to inconsistent behavior in conditional tests built on top-level variables defined as strings. Ansible 2.8 began changing this behavior. For example, if you set two conditions like this::
+
```
better to use a code-block here if we ever translate content.

acozine replied:
addressed by 708c3f226fefec7486a5022a182269bc6f4aa07d

samccann commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -62,12 +63,94 @@ Beginning in version 2.8, Ansible will warn if a module expects a string, but a
+* In Ansible 2.8 and later, you have the option of disabling conditional bare variables, so ``when: teardown`` always evaluates as ``True`` and ``when: not teardown`` always evaluates as ``False`` when ``teardown`` is a non-empty string (including ``'true'`` or ``'false'``)
+
+Ultimately, ``when: 'string'`` will always evaluate as ``True`` and ``when: not 'string'`` will always evaluate as ``False``, even if the value of ``'string'`` itself looks like a boolean. For users with playbooks that depend on the old behavior, we added a config setting that preserves it. You can use the ``ANSIBLE_CONDITIONAL_BARE_VARS`` environment variable or ``conditional_bare_variables`` in ...
+
```
consider putting the last two sentences in a note so users pay attention to the change coming in later releases.

acozine replied:
addressed by 708c3f226fefec7486a5022a182269bc6f4aa07d

samccann commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -62,12 +63,94 @@ Beginning in version 2.8, Ansible will warn if a module expects a string, but a
+^^^^^^^^^^^^^^^^^^^^^^^
+
+To prepare your playbooks for the new behavior, you must update your conditional statements so they accept only boolean values. For variables, you can use the ``bool`` filter to evaluate the string ``'false'`` as ``False``::
+
```
same comment - use code-blocks here and througout.

acozine replied:
addressed by 708c3f226fefec7486a5022a182269bc6f4aa07d

felixfontein commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -62,12 +63,94 @@ Beginning in version 2.8, Ansible will warn if a module expects a string, but a
+Nested variables
+^^^^^^^^^^^^^^^^
+
+The ``conditional_bare_variables`` setting does not affect nested variables. Subkey strings are already respected and not treated as booleans: ``when: complex_variable['subkey']`` is always ``True`` and ``when: not complex_variable['subkey']`` is always ``False``. If you want a string subkey to be evaluated as a boolean you must use the ``bool`` filter.
```
I think this should be a bit more explicit. When reading it, it took me quite a while to understand that "subkey strings" did not mean the `'subkey'` string, but that `complex_variable['subkey']` is a string. (Also, if it is the empty string, it will evaluate as `False`.)

acozine replied:
is this addressed by 708c3f226fefec7486a5022a182269bc6f4aa07d?

acozine replied:
grr, with 897b805ff71b6d5f222dd089c83932edc95f7c1f to address the empty string case?

felixfontein commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -62,12 +63,94 @@ Beginning in version 2.8, Ansible will warn if a module expects a string, but a
+Nested variables
+^^^^^^^^^^^^^^^^
+
+The ``conditional_bare_variables`` setting does not affect nested variables. Subkey strings are already respected and not treated as booleans: ``when: complex_variable['subkey']`` is always ``True`` and ``when: not complex_variable['subkey']`` is always ``False``. If you want a string subkey to be evaluated as a boolean you must use the ``bool`` filter.
```
I think this should be a bit more explicit. When reading it, it took me quite a while to understand that "subkey strings" did not mean the `'subkey'` string, but that `complex_variable['subkey']` is a string. (Also, if it is the empty string, it will evaluate as `False`.)

acozine committed:
incorporate PR feedback

acozine committed:
adds note about empty strings

felixfontein commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -62,12 +63,108 @@ Beginning in version 2.8, Ansible will warn if a module expects a string, but a
+
+To double-interpolate variable values, use curly braces:
+
+.. code-block:: yaml
```
```suggestion
.. code-block:: yaml+jinja
```

felixfontein commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -62,12 +63,108 @@ Beginning in version 2.8, Ansible will warn if a module expects a string, but a
+Nested variables
+^^^^^^^^^^^^^^^^
+
+The ``conditional_bare_variables`` setting does not affect nested variables. Any string value assigned to a subkey is already respected and not treated as a boolean: ``when: complex_variable['any non-empty string']`` is always ``True`` and ``when: not complex_variable['any non-empty string']`` is always ``False``. If you want a string subkey like ``complex_variable['false']`` to be evaluated as a ...
```
Maybe this instead:
```suggestion
The ``conditional_bare_variables`` setting does not affect nested variables. Any string value assigned to a subkey is already respected and not treated as a boolean: in case "complex_variable['subkey']" is a non-empty string, ``when: complex_variable['subkey']`` is always ``True`` and ``when: not complex_variable['subkey']`` is always ``False``. If you want a string subkey like ``complex_variable['subkey']`` (with value `'true'` or `'false'`) to be evaluated as a boolean you must use the ``bool`` filter.
```

acozine committed:
more PR feedback

acozine committed:
formatting mistake

felixfontein reviewed (approved):
LGTM

acozine committed:
flesh out note
